machineconfiguration/v1alpha1/types_machineosconfig.go

```diff
@@ -0,0 +1,233 @@
+
+// MachineOSConfigList describes all configurations for image builds on the system
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineOSConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []MachineOSConfig `json:"items"`
+}
+
+// MachineOSConfigSpec describes user-configurable options as well as information about a build process.
+type MachineOSConfigSpec struct {
+	// machineConfigPool is the pool which the build is for
+	// +kubebuilder:validation:Required
+	MachineConfigPool MachineConfigPoolReference `json:"machineConfigPool"`
+	// buildSettings is where user input options for the build live
+	// +kubebuilder:validation:Required
+	BuildSettings BuildSettings `json:"buildSettings"`
+	// +kubebuilder:validation:Required
+	// rebuild allows the user to manually trigger a rebuild
+	Rebuild MachineOSRebuild `json:"rebuild"`
+}
+
+// MachineOSConfigStatus describes the status this config object and relates it to the builds associated with this MachineOSConfig
+type MachineOSConfigStatus struct {
+	// observedGeneration represents the generation observed by the controller.
+	// this field is updated when the user changes the configuration in BuildSettings or the MCP this object is associated with.
+	// +kubebuilder:validation:Required
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	// buildHistory contains previous iterations of failed, interrupted, or succeded builds related to this config.
+	// this will only detail the 3 previous builds associated with the MachineOSConfig.
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=3
+	// +optional
+	BuildHistory []BuildHistory `json:"buildHistory,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
+	// finalImagePullspec is the fully qualified image pull spec used by the MCO to pull down the new OSImage. This must include sha256.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=447
+	// +kubebuilder:validation:XValidation:rule=`(self.split('@').size() == 2 && self.split('@')[1].matches('^sha256:[a-f0-9]{64}$'))`,message="the OCI Image reference must end with a valid '@sha256:<digest>' suffix, where '<digest>' is 64 characters long"
+	// +kubebuilder:validation:XValidation:rule=`(self.split('@')[0].matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?/([a-zA-Z0-9-_]{0,61}/)?[a-zA-Z0-9-_.]*?$') || self.matches('^[^.]+\\.[^.]+\\.svc:\\d+\\/[^\\/]+\\/[^\\/]+:[^\\/]+$'))`,message="the OCI Image name should follow the host[:port][/namespace]/name format, resembling a valid URL without the scheme"
+	// +optional
+	FinalImagePullspec string `json:"finalImagePullspec"`
```

Comment:
from an API consumer perspective, this is the `currentImagePullSpec`, right?

Reply:
this is the version with the sha256. but, since what I think you're saying is that this is on the config level and can change, then yeah I can rename this current rather than final